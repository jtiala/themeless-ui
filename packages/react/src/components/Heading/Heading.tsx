import { cn, CommonComponentProps } from "@themeless-ui/utils";
import { createElement, ReactNode } from "react";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

export type HeadingProps = {
  /**
   * The level of the heading.
   */
  level: HeadingLevel;

  /**
   * Heading content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const className = cn("heading");

/**
 * Display a section heading.
 */
export function Heading({ level, children, testId }: HeadingProps) {
  const element = "h" + level;

  return createElement(element, { className, "data-testid": testId }, children);
}
