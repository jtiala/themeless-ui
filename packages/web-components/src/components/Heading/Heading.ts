/* eslint-disable lit/no-invalid-html */
/* eslint-disable lit/binding-positions */
import style from "@themeless-ui/style/dist/heading.css?inline";
import { cn } from "@themeless-ui/utils";
import { unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import { html, unsafeStatic } from "lit/static-html.js";
import { TUIComponent } from "../../utils";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

const className = cn("heading");

@customElement("tui-heading")
export class Heading extends TUIComponent {
  /**
   * The level of the heading.
   */
  @property({ type: Number })
  level!: HeadingLevel;

  static styles = unsafeCSS(style);

  override render() {
    const tag = unsafeStatic(`h${this.level}`);

    return html`
      <${tag} class="${className}" id="${this.id}" data-testid="${this.testId}">
        <slot></slot>
      </${tag}>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-heading": Heading;
  }
}
