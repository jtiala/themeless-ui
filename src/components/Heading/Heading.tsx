import { createElement, ReactNode } from "react";
import { CommonComponentProps } from "../../utils/components";
import { cn } from "../../utils/css";

type HeadingLevel = 1 | 2 | 3 | 4 | 5 | 6;

export type HeadingProps = {
  /**
   * The level of the heading.
   */
  level: HeadingLevel;

  /**
   * Heading content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const className = cn("heading");

/**
 * Display a section heading.
 */
export function Heading({ level, children, testId, ...rest }: HeadingProps) {
  const element = "h" + level;

  return createElement(
    element,
    { className, ...rest, "data-testid": testId },
    children
  );
}
