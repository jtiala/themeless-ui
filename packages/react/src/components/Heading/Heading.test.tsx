import {
  cleanup,
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
} from "../../../test";
import { Heading } from "./Heading";

const defaultHeading = (
  <Heading level={1} testId="heading">
    Lorem ipsum dolor sit amet.
  </Heading>
);

describe("Heading", async () => {
  it("should render the heading", () => {
    expect(componentToJson(renderer.create(defaultHeading))).toMatchSnapshot();

    render(defaultHeading);

    const heading = screen.getByTestId("heading");

    expect(heading).toBeInTheDocument();
    expect(heading).toHaveTextContent("Lorem ipsum dolor sit amet.");
  });

  it("should render correct html element", () => {
    const levels = [1, 2, 3, 4, 5, 6] as const;

    levels.forEach((level) => {
      cleanup();

      const element = (
        <Heading level={level} testId="heading">
          Heading {level}
        </Heading>
      );

      expect(componentToJson(renderer.create(element))).toMatchSnapshot();

      render(element);

      const heading = screen.getByTestId("heading");
      expect(heading).toBeInTheDocument();
      expect(heading).toHaveTextContent(`Heading ${level}`);
      expect(heading.tagName.toLowerCase()).toBe(`h${level}`);
    });
  });
});
