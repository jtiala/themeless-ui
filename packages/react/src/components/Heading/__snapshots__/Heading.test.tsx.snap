// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Heading > should render correct html element 1`] = `
<h1
  className="tui__heading"
  data-testid="heading"
>
  Heading 
  1
</h1>
`;

exports[`Heading > should render correct html element 2`] = `
<h2
  className="tui__heading"
  data-testid="heading"
>
  Heading 
  2
</h2>
`;

exports[`Heading > should render correct html element 3`] = `
<h3
  className="tui__heading"
  data-testid="heading"
>
  Heading 
  3
</h3>
`;

exports[`Heading > should render correct html element 4`] = `
<h4
  className="tui__heading"
  data-testid="heading"
>
  Heading 
  4
</h4>
`;

exports[`Heading > should render correct html element 5`] = `
<h5
  className="tui__heading"
  data-testid="heading"
>
  Heading 
  5
</h5>
`;

exports[`Heading > should render correct html element 6`] = `
<h6
  className="tui__heading"
  data-testid="heading"
>
  Heading 
  6
</h6>
`;

exports[`Heading > should render the heading 1`] = `
<h1
  className="tui__heading"
  data-testid="heading"
>
  Lorem ipsum dolor sit amet.
</h1>
`;
